Frameworks/Foundation/NSOperation.mm

```diff
@@ -1,115 +1,112 @@
 
-    int executing : 1;
-    int cancelled : 1;
-    int finished : 1;
+/**
+ @Status Interoperable
+*/
++ (BOOL)automaticallyNotifiesObserversForKey:(NSString*)key {
+    // This class dispatches its own notifications.
+    return NO;
+}
 
-    NSOperationPriv() {
-        memset(this, 0, sizeof(NSOperationPriv));
-        pthread_cond_init(&finishCondition, 0);
+- (void)observeValueForKeyPath:(NSString*)keyPath ofObject:(id)object change:(NSDictionary*)change context:(void*)context {
+    if ([keyPath isEqualToString:@"isFinished"]) {
+        pthread_mutex_lock(&_finishLock);
+        pthread_cond_broadcast(&_finishCondition);
+        pthread_mutex_unlock(&_finishLock);
+    }
+}
 
+- (id)init {
+    if (self = [super init]) {
+        _dependencies.attach([NSMutableArray new]);
+        pthread_cond_init(&_finishCondition, 0);
         pthread_mutexattr_t attr;
-
         pthread_mutexattr_init(&attr);
         pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
-        pthread_mutex_init(&finishLock, &attr);
+        pthread_mutex_init(&_finishLock, &attr);
         pthread_mutexattr_destroy(&attr);
-    }
+        pthread_mutex_init(&_dependenciesLock, NULL);
+        pthread_mutex_init(&_completionBlockLock, NULL);
 
-    ~NSOperationPriv() {
-        pthread_cond_destroy(&finishCondition);
-        pthread_mutex_destroy(&finishLock);
+        [self addObserver:self forKeyPath:@"isFinished" options:0 context:NULL];
     }
-};
-#else
-struct NSOperationPriv;
-#endif
 
-@interface NSOperation () {
-    struct NSOperationPriv* priv;
+    return self;
 }
-@end
-
-@implementation NSOperation
 
 /**
  @Status Interoperable
 */
-+ (BOOL)automaticallyNotifiesObserversForKey:(NSString*)key {
-    // This class dispatches its own notifications.
-    return NO;
+- (void)addDependency:(id)operation {
+    pthread_mutex_lock(&_dependenciesLock);
+    [_dependencies addObject:operation];
+    pthread_mutex_unlock(&_dependenciesLock);
 }
 
 /**
  @Status Interoperable
 */
-+ (id)allocWithZone:(NSZone*)zone {
-    NSOperation* ret = [super allocWithZone:zone];
-
-    ret->priv = new NSOperationPriv();
-
-    return ret;
+- (void)removeDependency:(NSOperation*)operation {
+    pthread_mutex_lock(&_dependenciesLock);
+    [_dependencies removeObject:operation];
+    pthread_mutex_unlock(&_dependenciesLock);
 }
 
 /**
  @Status Interoperable
 */
-- (NSOperationQueuePriority)queuePriority {
-    return priv->priority;
-}
+- (BOOL)isReady {
+    if ([self isCancelled]) {
+        return YES;
+    }
 
-/**
- @Status Interoperable
-*/
-- (void)addDependency:(id)operation {
-    if (priv->dependencies == nil) {
-        priv->dependencies = [[NSMutableArray alloc] init];
+    pthread_mutex_lock(&_dependenciesLock);
+    int count = [_dependencies count];
+
+    for (int i = 0; i < count; i++) {
+        id op = [_dependencies objectAtIndex:i];
+
+        if (![op isFinished]) {
+            pthread_mutex_unlock(&_dependenciesLock);
```

Comment:
let us use std:: functions here and make the code a bit simpler to maintain with RAII